common/nexus/incoming_service_registry.go

```diff
@@ -0,0 +1,349 @@
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"go.temporal.io/api/nexus/v1"
+	"go.temporal.io/api/serviceerror"
+
+	"go.temporal.io/server/api/matchingservice/v1"
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/backoff"
+	"go.temporal.io/server/common/dynamicconfig"
+	"go.temporal.io/server/common/future"
+	"go.temporal.io/server/common/headers"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	p "go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/internal/goro"
+)
+
+type (
+	IncomingServiceRegistryConfig struct {
+		refreshLongPollTimeout dynamicconfig.DurationPropertyFn
+		refreshPageSize        dynamicconfig.IntPropertyFn
+		refreshMinWait         dynamicconfig.DurationPropertyFn
+		refreshRetryPolicy     backoff.RetryPolicy
+	}
+
+	// IncomingServiceRegistry manages a cached view of Nexus incoming services.
+	// Services are lazily-loaded into memory on the first read. Thereafter, service data is kept up to date by
+	// background long polling on matching service ListNexusIncomingServices.
+	IncomingServiceRegistry struct {
+		config *IncomingServiceRegistryConfig
+
+		status           int32
+		serviceDataReady *future.FutureImpl[struct{}]
+
+		dataLock     sync.RWMutex // protects tableVersion and services
+		tableVersion int64
+		services     map[string]*nexus.IncomingService // mapping of service ID -> service
+
+		refreshPoller *goro.Handle
+
+		matchingClient matchingservice.MatchingServiceClient
+		persistence    p.NexusIncomingServiceManager
+		logger         log.Logger
+	}
+)
+
+func NewIncomingServiceRegistryConfig(dc *dynamicconfig.Collection) *IncomingServiceRegistryConfig {
+	config := &IncomingServiceRegistryConfig{
+		refreshLongPollTimeout: dc.GetDurationProperty(dynamicconfig.FrontendRefreshNexusIncomingServicesLongPollTimeout, 5*time.Minute),
+		refreshPageSize:        dc.GetIntProperty(dynamicconfig.NexusIncomingServiceListDefaultPageSize, 1000),
+		refreshMinWait:         dc.GetDurationProperty(dynamicconfig.FrontendRefreshNexusIncomingServicesLongPollTimeout, 1*time.Second),
+	}
+	config.refreshRetryPolicy = backoff.NewExponentialRetryPolicy(config.refreshMinWait()).WithMaximumInterval(config.refreshLongPollTimeout())
+	return config
+}
+
+func NewIncomingServiceRegistry(
+	config *IncomingServiceRegistryConfig,
+	matchingClient matchingservice.MatchingServiceClient,
+	persistence p.NexusIncomingServiceManager,
+	logger log.Logger,
+) *IncomingServiceRegistry {
+	return &IncomingServiceRegistry{
+		status:           common.DaemonStatusInitialized,
+		config:           config,
+		serviceDataReady: future.NewFuture[struct{}](),
+		services:         make(map[string]*nexus.IncomingService),
+		matchingClient:   matchingClient,
+		persistence:      persistence,
+		logger:           logger,
+	}
+}
+
+func (r *IncomingServiceRegistry) Start() {
+	if !atomic.CompareAndSwapInt32(&r.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
```

Comment:
We should get rid of this pattern entirely, but at least stop using it for new code. The component should be started and stopped by fx lifecycle hooks and then we never have to worry about serializing start/stop.